internal/services/market.go

```diff
@@ -1,18 +1,104 @@
 type Market struct {
 	stor MarketPlaceholder
 }
 
 type MarketPlaceholder interface {
-	GetOrder()
+	GetOrders(ctx context.Context, userID *uuid.UUID) ([]model.Order, error)
+	SetOrder(ctx context.Context, userID *uuid.UUID, order string, isPreorder bool, withdraw *decimal.Decimal) error
+	IsExistForUser(ctx context.Context, userID *uuid.UUID, order string) (isExist bool, err error)
+	IsExistForOtherUsers(ctx context.Context, userID *uuid.UUID, order string) (isExist bool, err error)
+	GetAccruals(ctx context.Context, userID *uuid.UUID) (accrual *decimal.Decimal, err error)
+	GetWithdrawns(ctx context.Context, userID *uuid.UUID) (withdrawn *decimal.Decimal, err error)
+	Withdrawals(ctx context.Context, userID *uuid.UUID) ([]model.Withdrawals, error)
+	IsPreOrder(ctx context.Context, userID *uuid.UUID, order string) (isPreOrder bool, err error)
+	MovePreOrder(ctx context.Context, order *model.Order) (err error)
 }
 
 func NewMarket(stor MarketPlaceholder) *Market {
-	return &Market{}
+	return &Market{stor: stor}
+}
+
+func (m *Market) SetOrder(ctx context.Context, isPreOrder bool, order *model.Order) (existed bool, err error) {
+	// Add order to the database.
+	err = m.stor.SetOrder(ctx, order.UserID, order.Onumber, isPreOrder, order.Bonus.Used)
+	if err != nil {
+		var pgErr *pgconn.PgError
+		// If Order exist in the DataBase
+		if errors.As(err, &pgErr) && pgerrcode.UniqueViolation == pgErr.Code {
+			zap.S().Infoln("Order exist: ", order.Onumber)
+			return true, err
+		}
+		zap.S().Errorln("Set order error: ", order)
+	}
+
+	return false, nil
+}
+
+func (m *Market) GetOrders(ctx context.Context, userID *uuid.UUID) (orders []model.Order, err error) {
+
+	orders, err = m.stor.GetOrders(ctx, userID)
+	if err != nil {
+		return nil, err
+	}
+	return orders, nil
+}
+
+func (m *Market) IsPreOrder(ctx context.Context, userID *uuid.UUID, order string) (isPreOrder bool, err error) {
+	return m.stor.IsPreOrder(ctx, userID, order)
 }
 
-func (r *Market) GetOrder() {
+func (m *Market) MovePreOrder(ctx context.Context, order *model.Order) (err error) {
+	return m.stor.MovePreOrder(ctx, order)
+}
+
+func (m *Market) IsExistForUser(ctx context.Context, userID *uuid.UUID, order string) (isExist bool, err error) {
+	return m.stor.IsExistForUser(ctx, userID, order)
+}
+
+func (m *Market) IsExistForOtherUsers(ctx context.Context, userID *uuid.UUID, order string) (isExist bool, err error) {
+	return m.stor.IsExistForOtherUsers(ctx, userID, order)
+}
+
+func (m *Market) GetBalance(ctx context.Context, userID *uuid.UUID) (acc *decimal.Decimal, withdrawn *decimal.Decimal, err error) {
+	acc, err = m.stor.GetAccruals(ctx, userID)
+	if err != nil {
+		return nil, nil, err
+	}
+	withdrawn, err = m.stor.GetWithdrawns(ctx, userID)
+	if err != nil {
+		return nil, nil, err
+	}
+	return
+}
+
+func (m *Market) CheckBalance(ctx context.Context, userID *uuid.UUID, amount *decimal.Decimal) (isEnough bool, err error) {
+	acc, err := m.stor.GetAccruals(ctx, userID)
```

Comment:
по логике тут нужен баланс, а не только начисления. Т.к. денег с учетом списаний может быть меньше
```suggestion
	acc, err := m.stor.GetBalance(ctx, userID)
```

Reply:
Поправил логигу расчета бонусов, сделал отдельное поле в таблице пользователя и счтитаю все в нем.